server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,650 @@
+                        prefix_file = f"{prefix_file}.prefix"
+                        mdlogcfg = PbenchMDLogConfig(
+                            os.path.join(incoming_path, "metadata.log")
+                        )
+                        prefix = ""
+                        try:
+                            prefix = mdlogcfg.get("run", "prefix")
+                        except NoSectionError:
+                            pass
+                        except NoOptionError:
+                            pass
+                        if prefix_path == "":
+                            if prefix:
+                                rhier.add_unexpected_controllers(
+                                    "bad_prefixes", controller, path
+                                )
+                            elif os.path.exists(prefix_file):
+                                rhier.add_unexpected_controllers(
+                                    "unused_prefix_files", controller, path
+                                )
+                        else:
+                            if prefix:
+                                if prefix != prefix_path:
+                                    rhier.add_unexpected_controllers(
+                                        "bad_prefixes", controller, path
+                                    )
+                            elif not os.path.exists(prefix_file):
+                                rhier.add_unexpected_controllers(
+                                    "missing_prefix_files", controller, path
+                                )
+                            else:
+                                f = 0
+                                try:
+                                    with open(prefix_file, "r") as file:
+                                        prefix = file.read().replace("\n", "")
+                                except Exception:
+                                    f = 1
+                                if f == 1:
+                                    rhier.add_unexpected_controllers(
+                                        "bad_prefix_files", controller, path
+                                    )
+                                else:
+                                    if prefix != prefix_path:
+                                        rhier.add_unexpected_controllers(
+                                            "bad_prefixes", controller, path
+                                        )
+                        if user_arg:
+                            """We are reviewing a user tree, so check the user in
+                            the configuration.  Version 002 agents use the
+                            metadata log to store a user as well.
+                            """
+                            verify_user_arg(
+                                rhier, mdlogcfg, user_arg, user_controller, path
+                            )
+
+    return
+
+
+# Controller Hierarchy
+hierarchy_dispatcher = {
```

Comment:
I don't know how many places we might want to re-use the Hierarchy classes, or how many might have to deal with "raw" incoming hierarchies; but in general it bothers me to have a class that's so tightly coupled with its user that we rely on the caller to run all this validation outside the class logic.

The point of a class is to be encapsulated and easily reusable, so much of this coupled external logic seems to defeat the purpose of the refactoring. Certainly it's easy to get away with it with a single caller, and maybe it doesn't matter until/unless we want to use Hierarchy somewhere else... but isn't the promise of reuse ultimately the point?

Perhaps I just don't understand enough of the history and context; and while I think I grasp the general purpose and content of the "controller" directory hierarchies, I certainly don't have any real experience with them.

Reply:
We can do one thing. Mainly verifylist of CONTROLLER hierarchy is the one which leads to verify_incoming and verify_results. We can return verifylist to the main and call them(verify_incoming and verify_results) with verifylist as a function argument.

Reply:
it will also lose hierarchy_dispatcher and verify_hierarchy function